### PR TITLE
 Enable rule and measure opt-outs

### LINKED ISSUE #13: Enable "not applicable" where acceptable
Some measures are not appropriate for all projects. For example, `Corvus.Monitoring` does not have any deployment scripts because it's something that is only ever deployed as a component of something else. It does not require any artifacts up in Azure or in a container, so there's nothing to do here.

But currently we are nontheless required to provide a score for the `Deployment` rule. If we do not provide a value for this, we lose points in the overall IMM score, which is misleading because there's nothing we could do that would improve this library in this respect. Equally it would be misleading just to score ourselves full marks on this area when we do nothing at all related to deployment in this project.

So we think it should be possible for projects to opt out of this measure. There are other measures for which this may also be appropriate (e.g., Insights and Telemetry are probably not something you want in a string manipulation library).

Not all measures should be optional, but it should be possible to define some as optional.

Additionally, for cumulative measures, individual elements might be optional. For example, the `Design & Implementation Documentation` measure includes points for "Logical, infrastructure, security & ops views" which might not be applicable to all library code. So we would want it to be possible to score full marks without checking all boxes in those cases, but without needing to mark the entire measure as not applicable.

### PR DESCRIPTION
**Pull request by idg10**
Rule definitions are mandatory by default, but the rule set may choose to define some rules as optional, or individual measures within a particular rule as optional. In these cases, projects may declare that they these rules or measures do not apply.

Resolves #13 

This also improves the precision of the domain model in some important respects. We now distinguish between the _definitions_ of rules and measures (as they appear in the master rule set), and specific _assertions_ of those rules and measures (as they appear in individual projects' `imm.yaml` files). This had already been causing problems as a result of the difference between age-based rule definitions and their corresponding entries in `imm.yaml`, but these new features add further visible differences. In any case, the use of one type to mean two completely different things was making the codebase hard to follow—you could never tell in any particular part of the code whether it was working with a master measure definition or some entry from a particular project's `imm.yaml`.

Also in this vein, we've got specialized measure definition and assertion types for age- and framework-related measures. This means we no longer have the nasty situation of a single class with a bunch of differnt properties, where only certain property combinations are valid. This required some extra work in the YAML deserialization to ensure that the correct types get instantiated.

### CODE REVIEW

File: Solutions/Endjin.Imm.Domain/Processing/NullRuleCalculator.cs [hunk 1]
```diff
@@ -5,12 +5,12 @@
 
     public class NullRuleCalculator : IRuleCalculator
     {
-        public decimal Percentage(Rule rule, IEvaluationContext context)
+        public decimal Percentage(RuleAssertion rule, IEvaluationContext context)
```
Comment (HowardvanRooijen) [inline Solutions/Endjin.Imm.Domain/Processing/NullRuleCalculator.cs]: Should the parameter name (and I suspect the interface definition) be updated too?

File: Solutions/Endjin.Imm.Domain/Repository/RuleDefinitionRepository.cs [hunk 2]
```diff
@@ -16,7 +16,7 @@ public RuleDefinitionRepository(IpMaturityMatrixRuleset ruleSet)
             this.ruleSet = ruleSet;
         }
 
-        public RuleDefinition GetDefinitionFor(Rule rule)
+        public RuleDefinition GetDefinitionFor(RuleAssertion rule)
```
Comment (HowardvanRooijen) [inline Solutions/Endjin.Imm.Domain/Repository/RuleDefinitionRepository.cs]: Should the parameter name be updated too?

File: Solutions/Endjin.Imm.Domain/Processing/NullRuleCalculator.cs (hunk 1 above)
Comment (idg10) [reply Solutions/Endjin.Imm.Domain/Processing/NullRuleCalculator.cs]: Good point. Fixed.

File: Solutions/Endjin.Imm.Domain/Repository/RuleDefinitionRepository.cs (hunk 2 above)
Comment (idg10) [reply Solutions/Endjin.Imm.Domain/Repository/RuleDefinitionRepository.cs]: Fixed - thanks!
